app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt

```diff
@@ -18,24 +18,25 @@ import androidx.compose.ui.platform.LocalContext
 import androidx.compose.ui.res.stringResource
 import androidx.compose.ui.tooling.preview.Preview
 import androidx.compose.ui.unit.Dp
-import coil.compose.rememberAsyncImagePainter
+import androidx.hilt.navigation.compose.hiltViewModel
 import com.wire.android.R
+import com.wire.android.model.UserAvatarAsset
 import com.wire.android.model.UserStatus
 import com.wire.android.ui.theme.wireColorScheme
 import com.wire.android.ui.theme.wireDimensions
 import com.wire.android.util.getUriFromDrawable
-import com.wire.android.util.toBitmap
 
 @Composable
 fun UserProfileAvatar(
-// TODO: in the future we would not pass here null, but while developing there may be users having no avatar asset
-    avatarAssetByteArray: ByteArray? = null,
+    userAvatarAsset: UserAvatarAsset? = null,
     status: UserStatus = UserStatus.NONE,
     isClickable: Boolean = false,
     size: Dp = MaterialTheme.wireDimensions.userAvatarDefaultSize,
     modifier: Modifier = Modifier,
     onClick: (() -> Unit)? = null
 ) {
+    val imageLoader = hiltViewModel<CurrentSessionViewModel>().wireSessionImageLoader
```

Comment:
nice workaround :P however I can see a problem with this, from what I know hiltViewModel() gets a ViewModel instance by scoping it to NavBackStackEntry, so if we use it in different places in our app, each composable() or NavHost will create its own instance of this ViewModel, which will create its own instance of WireSessionImageLoader().

If this work-around is the only way we can use this ImageLoader inside UserProfileAvatar, perhaps we can make WireSessionImageLoader a Singleton. 

I never did such things in Coil like you did here , which I am impressed by :P but by quickly looking into the docs it looks like you can set the ImageLoader instance with this : 

```
  @JvmStatic
    @Synchronized
    fun setImageLoader(imageLoader: ImageLoader) {
        this.imageLoaderFactory = null
        this.imageLoader = imageLoader
    }
```
    
it looks like for example this method : 
```

@Composable
fun rememberAsyncImagePainter(
    model: Any?,
    transform: (State) -> State = DefaultTransform,
    onState: ((State) -> Unit)? = null,
    contentScale: ContentScale = ContentScale.Fit,
    filterQuality: FilterQuality = DefaultFilterQuality,
) = rememberAsyncImagePainter(
    model = model,
    imageLoader = LocalImageLoader.current,
    transform = transform,
    onState = onState,
    contentScale = contentScale,
    filterQuality = filterQuality
)    
```

makes use of LocalImageLoader.current, which in return looks like takes this singleton instance from Coil which you can set with the method above :p 

what I want to say, maybe it is worth to investigate if we can do it so that we just set this ImageLoader once and then coil does the rest for us so that we can keep using rememberAsyncImagePainter() for example without the work-around above, like we can configure Retrofit for example by adding some interceptor to the instance of Retrofit 

if there is no way around it, I am also fine with this work-around, no such a big deal and if it works correctly why not just keep it, it does not bring that much of confusion

Reply:
The main issue is that we cannot have a generic static image loader.

It needs to be scoped down to the user/session level.

If we switch from one session to another, the image loader should change in order to be able to load the assets properly (different backends, authentication, or even private assets).

So, it's impossible to get away with a singleton. Unless we start managing the state of this singleton, and keep updating the instance on the go, which sounds a bit weird, manually managing these moving pieces on each session change.

Reply:
I added some explanation on the scope and the problems this tries to solve [here](https://github.com/wireapp/wire-android-reloaded/blob/app/src/main/kotlin/com/wire/android/util/ui/WireSessionImageLoader.kt/b03cf0dae88df723e6e2569df6d05f6784065975).

Reply:
One example:

User A is logged in, we set the static instance of the ImageLoader and it's able to load assets (private, public, etc) for User A.

User B logs in on the same app, what happens? I wanted to leave DI to handle this, and keep the behaviour consistent with all the other UseCases and ViewModels.
If we could have a custom scope, that we initiate and destroy on a Session basis, that can work too.

Reply:
It seems the link to the explanation on the scope is broken :/

Reply:
I do not think I have enough concept to understand what you are saying ;p but if DI let us inject those ImageLoaders provide the proper use case to them then it should not be a problem if the user relogs because he reopens the app and reinitializes the dependency injection

Reply:
from what I see all we want to do is pass the instance of GetAvatarAssetUseCase to the ImageLoader, in theory I think it should be possible to configure it for example inside the @Provides method since the Coil instance has a method that allows us to set this ImageLoader, but it is all a theory at this point ;P

Reply:
> It seems the link to the explanation on the scope is broken :/

I'm on mobile, it seems the App doesn't let me get a proper link, but I tried to link to the documentation on `WireSessionImageLoader`.

Reply:
> from what I see all we want to do is pass the instance of GetAvatarAssetUseCase to the ImageLoader, in theory I think it should be possible to configure it for example inside the @Provides method since the Coil instance has a method that allows us to set this ImageLoader, but it is all a theory at this point ;P

DI is not initialized again.

`Application#onCreate` won't be called again when we just switch from one user to another.

I think only the activity(es) are recreated. So we can scope the ImageLoader to the main Activity for example.


Still, I guess one improvement we can have in the future is some custom scope, where we can start/stop in order to manipulate DI.

Reply:
then again in theory we could inject for example the ImageLoader.Factory into the onCreate of Activityu and configure the Coil by setting the ImageLoader on the Coil instance or something like that, again this is all in theory do not know if it is possible by doing so, if not please stop me telling you bullshit :P

Reply:
Then we need to configure Coil in multiple places, DI and Activity for setting the instance.

I think we can just leave in DI. Together with everything else, instead of needing to actively care about changing the singleton instance.

We definitely need our own DI component/scope for the future, something we don't have yet.
Once we have that, we can have a single loader that is only recreated if the current user session changes.

Also, calling directly the coil functions is not great if we ever want to replace it with some other solution.

Reply:
I do not understand why we would need to configure the Coil in multiple places, you can just do it once on Activity creation, regards calling coil functions I do not really know why we would switch to other solutions hmmm if we do it is a matter of changing one method inside UserProfileAvatar

Reply:
I tried to do something like this : 

```
    @Inject
    lateinit var wireSessionImageLoader: WireSessionImageLoader

    val viewModel: WireActivityViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        installSplashScreen()
        super.onCreate(savedInstanceState)
        viewModel.handleDeepLink(intent)

        Coil.setImageLoader(wireSessionImageLoader.coilImageLoader)
```

but indeed there is an issue with some stuff scoped to the ViewModelComponent, so for now I think, you approach is the best we got

Reply:
Next week: custom dagger component/scope.